Spurious -Wuninitialized for monostate emplace at -Og or with asan since gcc 12.1

Upstream: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111944

```c++
#include <boost/variant2/variant.hpp>
int main() {
    boost::variant2::variant<boost::variant2::monostate, int> r;
    r.emplace<boost::variant2::monostate>();
}
```

gcc 12.1+, -Og or -O -fsanitize=address, -Wuninitialized:

```
In member function 'constexpr void boost::variant2::detail::variant_storage_impl<std::integral_constant<bool, true>, T1, T ...>::emplace_impl(boost::mp11::mp_true, boost::mp11::mp_size_t<I>, A&& ...) [with long unsigned int I = 1; A = {}; T1 = boost::variant2::detail::none; T = {boost::variant2::monostate, int}]',
    inlined from 'constexpr void boost::variant2::detail::variant_storage_impl<std::integral_constant<bool, true>, T1, T ...>::emplace(boost::mp11::mp_size_t<I>, A&& ...) [with long unsigned int I = 1; A = {}; T1 = boost::variant2::detail::none; T = {boost::variant2::monostate, int}]' at boost/variant2/variant.hpp:705:27,
    inlined from 'constexpr void boost::variant2::detail::variant_base_impl<true, true, T ...>::emplace_impl(boost::mp11::mp_true, A&& ...) [with long unsigned int J = 1; U = boost::variant2::monostate; A = {}; T = {boost::variant2::monostate, int}]' at boost/variant2/variant.hpp:902:20,
    inlined from 'constexpr void boost::variant2::detail::variant_base_impl<true, true, T ...>::emplace(A&& ...) [with long unsigned int I = 0; A = {}; T = {boost::variant2::monostate, int}]' at boost/variant2/variant.hpp:921:33,
    inlined from 'constexpr U& boost::variant2::variant<T>::emplace(A&& ...) [with U = boost::variant2::monostate; A = {}; E = void; T = {boost::variant2::monostate, int}]' at boost/variant2/variant.hpp:1690:49:
boost/variant2/variant.hpp:696:9: warning: '<anonymous>' is used uninitialized [-Wuninitialized]
  696 |         *this = variant_storage_impl( mp11::mp_size_t<I>(), std::forward<A>(a)... );
      |         ^
boost/variant2/variant.hpp: In member function 'constexpr U& boost::variant2::variant<T>::emplace(A&& ...) [with U = boost::variant2::monostate; A = {}; E = void; T = {boost::variant2::monostate, int}]':
boost/variant2/variant.hpp:696:17: note: '<anonymous>' declared here
  696 |         *this = variant_storage_impl( mp11::mp_size_t<I>(), std::forward<A>(a)... );
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This affects Boost.System `result<void>` (e.g. test/result_emplace.cpp).

You already have a suppression for -Wmaybe-uninitialized, do you want to add one for -Wuninitialized as well?

--- a/include/boost/variant2/variant.hpp
+++ b/include/boost/variant2/variant.hpp
@@ -689,8 +689,11 @@ template<class T1, class... T> union variant_storage_impl<mp11::mp_true, T1, T..
 # pragma GCC diagnostic push
 // False positive in at least GCC 7 and GCC 10 ASAN triggered by monostate (via result<void>)
 # pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#if __GNUC__ >= 12
+// False positive in at least GCC 12 and GCC 13 ASAN and -Og triggered by monostate (via result<void>)
+# pragma GCC diagnostic ignored "-Wuninitialized"
 #endif
-
+#endif
         *this = variant_storage_impl( mp11::mp_size_t<I>(), std::forward<A>(a)... );
 
 #if defined(BOOST_GCC) && (__GNUC__ >= 7)

